thread: remove recursive mutex usage

hzhou opened the pull request:
## Pull Request Description

The provision of recursive mutex usage results in an unclear understanding of whether certain code is inside or outside a critical section and it is difficult in places where we need to yield the lock and not sure how many layers of critical section it is under. The situation is made worse by the transition to granular locks due to potential multiple locking.

This PR adds an assertion to alert all usage of recursive locking and takes measures to eliminate such usage. The recursive locking mechanism is still left in-place. For one, it is not much of a performance concern; for two, the mechanism of maintaining lock owner and counting provides very convenient debugging assistance. For example, simply having the assertion to alert all places with bad mutex usage is very helpful.

Note: this is split from PR #4385.

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:mpich/authorship

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/warnings
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/async

hzhou commented on src/include/mpir_request.h:
```diff
@@ -503,15 +503,28 @@ MPL_STATIC_INLINE_PREFIX int MPIR_Request_has_wait_fn(MPIR_Request * request_ptr
+    int mpi_errno;
+    MPID_THREAD_CS_EXIT(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
+    MPID_THREAD_CS_EXIT(VCI, MPIR_THREAD_VCI_GLOBAL_MUTEX);
+    mpi_errno = (request_ptr->u.ureq.greq_fns->wait_fn) (1,
```
Those function pointers are only set during the creation function `MPI_Grequest_start`, before visible to other threads.

The async failures in ch4 will be fixed in the following PRs. This PR is a prerequisite.

hzhou commented:
@minsii MIn, could you take look at a few commits that are related to RMA -- in particular, https://github.com/pmodels/mpich/pull/4425/commits/8a070bdf66d796aea28ac194b042306d736ebb41, https://github.com/pmodels/mpich/pull/4425/commits/82070042561340fe54bd08f0ff42bf402ef764ee? Feel free to review the entire PR if you feel like it 😄

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -396,18 +396,48 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
+/* Optimized versions to avoid exessive locking/unlocking */
+/* FIXME: use inline function rather macros for cleaner semantics */
+
 #define MPIDIU_PROGRESS_WHILE(cond)         \
```
Can we use inline function here instead of leaving a FIXME comment?

hzhou replied:
I am trying to limit the changes in this PR. While it appears a lot of code change, it is still within the similar code design as original code. The `FIXME` is a reminder for further cleanup. `inline function` is just one idea. After the multi-VCI development, different solution may become better choice. For example, we may have better refactoring so the way RMA code calls progress may look different by then. That's why I would just leave as comment now and re-evaluate once the threading code settles.

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -396,18 +396,48 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
+/* This macro is refactored for original code that progress in a do-while loop
+ * TODO: Can we consolidate with previous macro? Double check the reasoning.
+ */
+#define MPIDIU_PROGRESS_DO_WHILE(cond) \
```
Can the TODO be resolved?

hzhou replied:
It's orthogonal to this PR. Besides, once the multi-vci code goes in, the code and solution here may become different. So postponing the cleanup for now.

minsii commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -396,18 +396,48 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
+#define MPIDIU_PROGRESS_DO_WHILE(cond) \
+    do {                                        \
+        MPID_THREAD_CS_EXIT(VCI, MPIDI_global.vci_lock); \
+        do {                          \
```
This question might be unrelated to this specific PR: why `MPID_THREAD_CS_EXIT` is called before `MPID_THREAD_CS_ENTER` in these routines? Should it be ENTER->do something->EXIT?

hzhou replied:
It's a "YIELD" 😄 .

The code here are already inside progress lock, and it calls progress again. So they need yield to avoid recursive locking. 

The main goal of this PR is to remove recursive locking.

minsii replied:
I see. Since the locks are move into a single place now can you please add a short comment to explain it? Just to make me not ask the same question again:-)

hzhou replied:
ok, will do.

hzhou replied:
Updated.

minsii commented:
The RMA portion looks OK to me.

hzhou commented:
test:mpich/warnings

hzhou commented:
@raffenet I have rebased the branch to current master and added a comment to address Min's comment. @raffenet Do you have other questions or concerns to discuss?

raffenet commented on src/glue/romio/glue_romio.c:
```diff
@@ -96,8 +96,6 @@ void MPIR_Ext_cs_exit(void)
 void MPIR_Ext_cs_yield(void)
 {
     /* TODO: check whether the progress engine is blocked */
-    MPID_THREAD_CS_YIELD(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```
Should this then be `MPL_thread_yield()` and squashed with "romio: let romio have its own thread-safety"?

hzhou replied:
Let's pull @roblatham00 in :).

Rob, there are only two place call `MPIR_Ext_cs_yield`. `ad_iread_coll.c` and `ad_iwrite_coll.c`, both inside a progress loop polling `Grequest`'s `poll_fn`. Is there a case that the `poll_fn` can only proceed when other threads of the same process gets to run?

raffenet replied:
Initially added in 6cae5cab8a6d3f99b1b2f0ba317fa9ed48b604bc.

roblatham00 replied:
the ad_i{read,write}_coll.c routines support nonblocking collectives.  I didn't write that ( Sangmin did) so I can't say off the top of my head what the yield is trying to do. 

I did write the `poll_fn`: the rules for what you can do in a poll_fn are not strictly defined -- we only have my EuroMPI paper attempting to do so.  My intention for the `poll_fn` is, as name suggests, to never block.   "read a flag from a register" and "make a system call that will return immediately" are the two use cases I ever had in mind.

hzhou replied:
I see. Now ROMIO does not take MPICH lock what-so-ever, it is not blocking other threads running mpich progress. I guess the concern here is in the user-thread (argobot) case, we may still need explicit yield. I guess adding `MPL_thread_yield()` here is OK.

hzhou replied:
@roblatham00 Thanks. The concern here is not whether the poll_fn will block; the concern is whether it will progress. So the question is whether it can progress without the participation of other threads. But I guess the answer is no, guessing from the reason that yield was added in the first place. I'll add `MPL_thread_yield()`, which currently is a noop on linux, a sleep(0) on mac and windows, and I am pondering on making it a couple hundred nanoseconds of `sched_yield()` out-of ch4 necessity. So if you are concerned that you can't afford a, say, 1us delay in the loop, let us know :)

hzhou replied:
I'll add back the lock yield as well. If it was concerned the other threads need progress, then the other threads must need able to call ROMIO functions.

raffenet commented on src/mpi/request/test.c:
```diff
@@ -121,6 +121,7 @@ int MPI_Test(MPI_Request * request, int *flag, MPI_Status * status)
     MPIR_ERRTEST_INITIALIZED_ORDIE();
 
     MPID_THREAD_CS_ENTER(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
+    MPID_THREAD_CS_ENTER(VCI, MPIR_THREAD_VCI_GLOBAL_MUTEX);
```
nitpick: this and the ch4 wait patch could probably be squashed

hzhou replied:
OK.

hzhou commented:
test:mpich/warnings
test:mpich/ch3/most
test:mpich/ch4/most

raffenet commented on src/include/mpir_ext.h.in:
```diff
@@ -68,9 +68,10 @@ extern int MPIR_Ext_dbg_romio_terse_enabled;
 extern int MPIR_Ext_dbg_romio_verbose_enabled;
 
-/* to be called early by ROMIO's initialization process in order to setup init-time
+/* to be called early by MPIR_Init_thread when ROMIO is included in order to setup init-time
```
Hmm, I'm not sure we can move this initialization inside MPICH. That would break users such as Open MPI, no?

hzhou replied:
Hehe, "break" is a strong word. Yes, for other users, they'll need to update their code to adapt to this change. A library that requires init/finalize is not a hacky idea, right? Users just need to know that they need to call init/finalize. I think it is better than the current lazy init approach, which is obscure and non-expandable. From what I see, the current way other users to use ROMIO is a hacky process anyway -- requiring their own `mpir_ext.h`. FWIW, after this patch, ROMIO is nearly independent from the rest of the MPICH, and with another patch, we probably can get rid of `mpir_ext.h` and make ROMIO independent package, which should greatly simplify how other users use it.

As for the impact, does open-mpi directly take romio from mpich repository?

hzhou replied:
I guess we can simply use static mutex initialization. Let me try that.

hzhou replied:
> 
> 
> I guess we can simply use static mutex initialization. Let me try that.

Oh, I see why we can't. We are using MPL wrappers. I guess we can't assume pthread mutex is always available even though in practice it does.  So --

* Option 1. abandon MPL wrappers and directly use pthread_mutex for `romio_mutex`, adding hard dependency on pthread even for THREAD_SINGLE build

* Option 2. use MPL wrappers. Then we can't do lazy init.

raffenet replied:
> Hehe, "break" is a strong word. Yes, for other users, they'll need to update their code to adapt to this change. A library that requires init/finalize is not a hacky idea, right?

It is a backwards compatibility "break". 

> Users just need to know that they need to call init/finalize. 

Then this needs additional documentation. Still, I think limiting the scope of changes to get this merged would be a good idea.

> As for the impact, does open-mpi directly take romio from mpich repository?

Yes.

hzhou replied:
Moved the mutex init to lazy init. There is no place to call finalize, so we'll just leave it like that. Should be ok for pthread mutex, but it's a small leak for argobots.

raffenet replied:
Thanks. It's not perfect, but I think it should work :).

hzhou commented:
test:mpich/ch3/most

hzhou commented:
test:mpich/warnings

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/warnings

hzhou commented:
@raffenet Fixed a warning -- could you review/re-approve?

raffenet commented:
Sure, just one more question I had when thinking about this. Will there be any issue when running this version of ROMIO with an older MPICH-based distribution? I'm thinking about the users who develop their own ROMIO code to run on a Cray system. Is there anything to be concerned about? I'm still approving, just thinking out loud.

hzhou commented:
> Will there be any issue when running this version of ROMIO with an older MPICH-based distribution? I'm thinking about the users who develop their own ROMIO code to run on a Cray system.

Users write their own code using ROMIO and linking with mpich with romio already linked? Wouldn't there be duplicated symbols because ROMIO being linked twice?

Or, are you thinking that user build romio stand-alone and linking with mpich that doesn't have romio? Since there is nothing romio in the mpich, I don't see an issue.
